classes/form/username_form.php

```diff
@@ -0,0 +1,111 @@
+use bmdisco_domain\brand_domain;
+use local_signin\util;
+use moodleform;
+
+defined('MOODLE_INTERNAL') || die;
+
+require_once "{$CFG->libdir}/formslib.php";
+
+class username_form extends moodleform {
+    /**
+     * @override \moodleform
+     */
+    public function definition() {
+        global $frm;
+
+        $mform  = $this->_form;
+
+        $username = '';
+        if (isset($frm) && isset($frm->username) && $frm->username) {
+            $username = $frm->username;
+        }
+        $user_attributes = array('placeholder'   => $this->lang_string('form_username_placeholder'),
+                                 'additionalcss' => $this->lang_string('form_username_button_class'),
+                                 'autofocus'     => '',
+                                 'value'         => $username);
+        $mform->addElement('text', 'username', $this->lang_string('form_username_label'), $user_attributes);
+        $mform->setType('username', PARAM_USERNAME);
+
+        $mform->addElement('html', sprintf('<div class="%s">', $this->lang_string('form_username_remusername_class')));
+        $mform->addElement('advcheckbox', 'rememberme', '', $this->lang_string('form_username_remusername_label'), array ('id' => 'check_rememberme'));
+        $mform->addElement('html', '</div>');
+
+        $submit_attributes = array('additionalcss' => $this->lang_string('form_username_button_class'));
+        $mform->addElement('submit', 'submitusername', $this->lang_string('form_username_button_label'), $submit_attributes);
+
+        $mform->addElement('html', sprintf('<div class="%s"><a href="%s">%s</a></div>',
+            $this->lang_string('form_userpass_forgot_class'),
+            new \moodle_url('/local/signin/forgot.php'),
+            $this->lang_string('form_username_forgot_label')));
+    }
+
+    function set_data($default_values)
+    {
+        parent::set_data($default_values); // TODO: Change the autogenerated stub
+    }
+
+    public function lang_string($id) {
+        return get_string($id, util::MOODLE_COMPONENT);
+    }
+
+    /**
+     * Validate the username field.
+     *
+     * @param array $data
+     * @param array $files
+     * @return array|null
+     */
+    public function validation($data, $files) {
+        global $CFG, $PAGE;
+
+        if (!array_key_exists('username', $data)) {
+            return array('username' => $this->lang_string('form_username_not_provided'));
+        }
+
+        $username = $data['username'];
+
+        if (strlen($username) == 0 || !static::active_user_exists($username)) {
+            return array('username' => $this->lang_string('form_username_not_found_valid'));
+        }
+
+        if ($username == 'guest') {
+            return;
+        }
+
+        if (class_exists('bmdisco_domain\brand_domain')) {
+            $correct_domain = brand_domain::get_user_default_domain_obj($username, true) ?
+                brand_domain::get_user_default_domain_obj($username, true)->domain : '';
```

Comment:
This is inefficient: it'll hit the database twice, once with the condition and again with the value if the condition returns a truthy value.

This method has also been renamed in `bmdisco_domain` and will throw a `dml_missing_record_exception` if no matching domain was found, so we'll want to catch that here and recover gracefully.